### PR TITLE
support_TIF_classweights_and_exclusion

### PR DESCRIPTION
**Pull request by minh-doan**
- Support parsing TIF
- Support class_weights in model.py
- Allow exclusion when collecting samples for training/testing sets

### CODE REVIEW

File: deepometry/commands/command_evaluate.py [hunk 1]
```diff
@@ -90,19 +110,26 @@ def _evaluate(x, y, batch_size, directory, name, verbose):
...
+
     x = numpy.empty((len(pathnames),) + _shape(pathnames[0]), dtype=numpy.uint8)
 
     y = numpy.empty((len(pathnames),), dtype=numpy.uint8)
 
     label_to_index = {label: index for index, label in enumerate(sorted(labels))}
 
     for index, pathname in enumerate(pathnames):
-        label = os.path.split(os.path.dirname(pathname))[-1]
+        if os.path.isfile(pathname) == True: # in case there is a mixture of directories and files
```
Comment (mcquin) [inline deepometry/commands/command_evaluate.py]: The `== True` in `if os.path.isfile(pathname) == True` is redundant. `if os.path.isfile(pathname)` is enough.

Comment (mcquin) [inline deepometry/commands/command_evaluate.py]: If the pathname is a directory, then it will not be loaded. Is this the intended behavior?

File: deepometry/commands/command_fit.py [hunk 2]
```diff
@@ -55,7 +58,14 @@
     "--verbose",
     is_flag=True
 )
-def command(input, batch_size, directory, epochs, name, validation_split, verbose):
+@click.option(
```
Comment (mcquin) [inline deepometry/commands/command_fit.py]: Please apply the comments from `command_evaluate` here, too.

File: deepometry/commands/command_fit.py [hunk 3]
```diff
@@ -55,7 +58,14 @@
...
+    "--exclusion",
+    default=None,
+    help="A comma-separated list of prefixes (string) specifying the files that needs to be helf off from the training dataset."
+         " E.g., \"'patient_A', 'patient_X'\". All numpy arrays will be collected for fitting if this flag is omitted."
+)
+
+def command(input, exclusion, batch_size, directory, epochs, name, validation_split, verbose):
```
Comment (mcquin) [inline deepometry/commands/command_fit.py]: Please apply the comments from `command_evaluate` here, too.

File: deepometry/commands/command_fit.py [hunk 4]
```diff
@@ -78,26 +88,34 @@ def command(input, batch_size, directory, epochs, name, validation_split, verbos
     model.fit(
         x,
         y,
+        class_weight = get_class_weights(y),
```
Comment (mcquin) [inline deepometry/commands/command_fit.py]: Unnecessary whitespace between keyword arguments: `class_weight=get_class_weights(y)`

File: deepometry/commands/command_parse.py [hunk 5]
```diff
@@ -58,37 +59,55 @@ def command(input, output, channels, image_size, verbose):
 
     label_directories = glob.glob(os.path.join(input_directory, "*"))
 
-    parsed_channels = None if channels is None else _parse_channels(channels)
+    # Check extension
+    pathnames = glob.glob(os.path.join(label_directories[0], "*"))
+        ## TO_DO:
+        # how to make sure the label_directories[0] is not non-folder files? e.g. .DS_Store
```
Comment (mcquin) [inline deepometry/commands/command_parse.py]: You can use `os.path.isdir` to filter out directories from non-directories.

File: deepometry/model.py [hunk 6]
```diff
@@ -101,7 +101,8 @@ def fit(self, x, y, batch_size=32, epochs=512, validation_split=0.2, verbose=0):
             "epochs": epochs,
             "steps_per_epoch": len(x_train) // batch_size,
             "validation_steps": len(x_valid) // batch_size,
-            "verbose": verbose
+            "verbose": verbose,
+            "class_weight" : class_weight
```
Comment (mcquin) [inline deepometry/model.py]: "class_weight": class_weight

File: deepometry/parse.py [hunk 7]
```diff
@@ -33,7 +37,7 @@ def parse(pathname, output_directory, size, channels=None):
     if ext == ".cif":
         return _parse_cif(pathname, output_directory, size, channels)
 
-    raise NotImplementedError("Unsupported file format: {}".format(ext))
+    raise NotImplementedError("Expected file format: {}".format(".CIF"))
```
Comment (mcquin) [inline deepometry/parse.py]: I envisioned `parse` being a top-level function which delegated to different parse functions based on the file extensions. E.g.,

```python
def parse(pathname, output_directory, size, channels=None):
    ext = os.path.splitext(pathname)[-1].lower()

    if ext == ".cif":
        return _parse_cif(pathname, output_directory, size, channels)

    if ext == ".npy":
        return _parse_npy(pathname, output_directory, size, channels) 

    raise NotImplementedError("Unsupported file format: {}".format(ext))
```

I see how it would be annoying to call parse for every .TIF image. Maybe the first argument (`pathname`) could take either a single file name or a list of file names.

File: deepometry/parse.py [hunk 8]
```diff
@@ -52,21 +56,80 @@ def _parse_cif(pathname, output_directory, size, channels):
...
+
+    def channel_regex(channels):
+        return ".*" + "Ch(" + "|".join(str(channel) for channel in channels) + ")"
+
+    regex = channel_regex(channels)
+
+    nested_filenames = []
+
+    for label in labels:
+#        print("Parsing directory: {}".format(label))
```
Comment (mcquin) [inline deepometry/parse.py]: This comment can be deleted.

File: deepometry/parse.py [hunk 9]
```diff
@@ -52,21 +56,80 @@ def _parse_cif(pathname, output_directory, size, channels):
...
+
+        src_dir = os.path.join(src, label)
+
+        filenames = glob.glob("{}/*.tif".format(src_dir))
```
Comment (mcquin) [inline deepometry/parse.py]: Use `os.path.join` to concatenate paths: `os.path.join(src_dir, "*.tif")`

**Comment by mcquin** [comment]
There are some good changes here, Minh. Nice work!

Let's work next on getting the existing tests passing:

- Any call to `model.fit` needs to have the `class_weights` parameter specified. Alternatively, we could provide a sane default. For example, if `class_weights` is none, then use equal weights (1?).
- The evaluate command doesn't run [due to this line](https://github.com/minh-doan/deepometry-1/blob/a95226ca93ed34dafc4e4ffb90f32df736b67957/deepometry/commands/command_evaluate.py#L89). I made some comments in the PR on how we can change it that should resolve the issue.
- The default for `--exclusion` should probably be `""` instead of `None`, or we should use an `if` to skip the exclusion filtering when a pattern isn't provided.
- The parse commands are failing with "Unsupported file format: .cif" errors. Maybe there's something missing in the extension check.

File: deepometry/commands/command_evaluate.py [hunk 10]
```diff
@@ -62,13 +77,18 @@ def command(input, batch_size, directory, name, verbose):
...
     pathnames = []
 
     for directory in directories:
-        subdirectories = glob.glob(os.path.join(directory, "*"))
+        subdirectories = sorted(glob.glob(os.path.join(directory, "*")))
+
+        # transform the files of the same label into directory
+        subdirectory_pathnames = [glob.glob("{}/*.npy".format(subdirectory)) for subdirectory in subdirectories ]
+
+        nsamples = max([len(pathnames) for pathnames in subdirectory_pathnames]) if nsamples is None else nsamples = nsamples
```
[the remark this answers is not in the document]
Comment (minh-doan) [reply deepometry/commands/command_evaluate.py]: So I just delete the whole line altogether is that correct?

File: deepometry/commands/command_evaluate.py (hunk 1 above)
Comment (minh-doan) [reply deepometry/commands/command_evaluate.py]: yes, I wanted to only collect the numpy arrays (files) here. Temporary/accidental directories should not be loaded

File: deepometry/commands/command_evaluate.py [hunk 11]
```diff
@@ -90,19 +110,26 @@ def _evaluate(x, y, batch_size, directory, name, verbose):
     return model.model.metrics_names, metrics
 
 
-def _load(pathnames, labels):
+def _load(pathnames, labels, exclusion):
+
+    print('Before exclusion: ',len(pathnames))
+    pathnames = [x for x in pathnames if exclusion not in x]
```
[the remark this answers is not in the document]
Comment (minh-doan) [reply deepometry/commands/command_evaluate.py]: nice catch!
I fix it with a loop, can you please check

File: deepometry/parse.py (hunk 9 above)
Comment (minh-doan) [reply deepometry/parse.py]: I don't need glob.glob when using `os.path.join(src_dir, "*.tif")`, right?

**Comment by mcquin** [comment]
Usurped by #22
